### PR TITLE
Add pagination and validation

### PR DESCRIPTION
**Pull request by laptou**
- Comments now appear paginated to user
- Server now validates comments (ex.: comments more than 50000 bytes are truncated, comments containing HTML are flagged, although the flagging has no effect yet).

Screenshot 2020-06-03 at 14 32 06 - Display 1

### CODE REVIEW

File: portfolio/src/main/client/src/components/sections/comments.ts [hunk 1]
```diff
@@ -11,44 +14,128 @@ export interface CommentInfo {
   downvotes: number;
 }
 
+export interface CommentPage {
```
Comment (jessicaslaughter) [inline portfolio/src/main/client/src/components/sections/comments.ts]: nit: add JSDoc comment

File: portfolio/src/main/client/src/components/sections/comments.ts [hunk 2]
```diff
@@ -11,44 +14,128 @@ export interface CommentInfo {
...
-      <li class="error">Failed to load comments. ${reloadLink}</li>
-    `);
+  } else {
+    listEl.append(CommentEmptyState());
+  }
+
+  return data;
+}
+
+// eslint-disable-next-line valid-jsdoc
```
Comment (jessicaslaughter) [inline portfolio/src/main/client/src/components/sections/comments.ts]: why is this needed?

File: portfolio/src/main/client/src/components/sections/comments.ts [hunk 3]
```diff
@@ -70,40 +157,84 @@ const Comment = (comment: CommentInfo): HTMLElement => {
...
 };
 
+const CommentEmptyState = (): HTMLElement => {
+  return htmlElement`
+  <li class="comment-empty-state">
+    You've reached the end. No more comments.
+  </li>`;
+};
+
 export const CommentSection = (): HTMLElement => {
```
Comment (jessicaslaughter) [inline portfolio/src/main/client/src/components/sections/comments.ts:216]: we missed this last time, but this export should also have a JSDoc comment

**Review by jessicaslaughter — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 4]
```diff
@@ -43,38 +48,89 @@
 public class CommentServlet extends HttpServlet {
   private static Gson gson = new Gson();
   private static DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  private static Pattern htmlDetector =
+      Pattern.compile("<\\w+(\\s*\\w+\\s*(=\\s*['\"].*['\"]))*>.*<\\/\\w+>", Pattern.DOTALL);
```
Comment (jpnickolas) [inline portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java:52]: https://stackoverflow.com/a/11230103

I'm pretty sure I can insert unclosed html tags, such as images (``). I could also just not close my html tags and see how those wreak havoc on your layout `<p><div><script>alert("dap the haters");` If there's a genuine html parser you can use, that's great. Otherwise, it's fine to have something that only works 90% of the time.

Comment (laptou) [reply portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java:52]: I'm fine with this only working 90% of the time, since I insert user-supplied data using `innerText` anyway. This is just for novelty, not security.

File: portfolio/src/main/client/src/components/sections/comments.ts (hunk 2 above)
Comment (laptou) [reply portfolio/src/main/client/src/components/sections/comments.ts]: It actually isn't, I removed it

File: portfolio/src/main/client/src/components/sections/comments.ts [hunk 5]
```diff
@@ -2,7 +2,10 @@ import {htmlElement} from '@src/util/html';
 import '@res/style/sections/comments.scss';
 import {LabeledInput} from '../controls/labeled-input';
 
-export interface CommentInfo {
+type Cursor = string;
+type CommentPageStack = [null, ...Cursor[]];
```
Comment (jpnickolas) [inline portfolio/src/main/client/src/components/sections/comments.ts]: Why does a page stack always start with null?

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 6]
```diff
@@ -43,38 +48,89 @@
...
+      String username = (String) e.getProperty("username");
+      String name = (String) e.getProperty("name");
+      String content = ((Text) e.getProperty("content")).getValue();
+      boolean shameful = (boolean) e.getProperty("shameful");
+      return new Comment(id, username, name, content, shameful);
+    }).collect(Collectors.toList());
+
+    JsonObject root = new JsonObject();
+    root.add("comments", gson.toJsonTree(comments));
+    root.addProperty("next", results.getCursor().toWebSafeString());
```
Comment (jpnickolas) [inline portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java]: git: give this a more descriptive name like "nextPageToken"

File: portfolio/src/main/client/src/components/sections/comments.ts (hunk 1 above)
Comment (jessicaslaughter) [reply portfolio/src/main/client/src/components/sections/comments.ts]: CommentPage interface still needs commenting

**Review by jessicaslaughter — CHANGES_REQUESTED** [review]

File: portfolio/src/main/client/src/components/sections/comments.ts [hunk 7]
```diff
@@ -14,44 +17,136 @@ export interface CommentInfo {
...
+ * @param pages The comment page stack.
+ * @param limit The number of comments per page.
+ */
+async function loadCurrentPage(
+  el: HTMLElement,
+  pages: CommentPageStack,
+  limit: number
+) {
+  // if pages.length is less than 2, then no page
+  // has been loaded yet
```
Comment (jpnickolas) [inline portfolio/src/main/client/src/components/sections/comments.ts]: This comment is unhelpful since it just describes what the following code does. You should instead explain its intentions. For example "The first cursor in the page stack is always null. We need at least one string cursor to load a page"

File: portfolio/src/main/client/src/components/sections/comments.ts [hunk 8]
```diff
@@ -14,44 +17,136 @@ export interface CommentInfo {
...
+  if (pages.length < 2) {
+    throw new Error('no comments page is currently loaded');
+  }
+
+  const result = await load(el, pages[pages.length - 2], limit);
```
Comment (jpnickolas) [inline portfolio/src/main/client/src/components/sections/comments.ts]: This array manipulation is pretty difficult to follow here. You should consider using a linked list instead. It's not intuitive to realize that this array is the content with nulls buffered on each side so they can be overwritten, and you can probably see this confusion in my previous comments.

```
interface Cursor {
  prev: Cursor|null;
  next: Cursor|null;
  value: string;
}
```

File: portfolio/src/main/client/src/components/sections/comments.ts [hunk 9]
```diff
@@ -14,44 +17,136 @@ export interface CommentInfo {
...
+ * @param commentsEl The element containing the comment list.
+ * @param formEl The form to submit.
+ * @param pages The comment page stack.
+ * @param limit The number of comments per page.
  */
-async function submitForm(el: HTMLElement, form: HTMLFormElement) {
+async function submitForm(
+  commentsEl: HTMLElement,
+  formEl: HTMLFormElement,
+  pages: [null, ...string[]],
```
Comment (jpnickolas) [inline portfolio/src/main/client/src/components/sections/comments.ts]: nit: the type of this should be CommentPageState

**Review by jpnickolas — CHANGES_REQUESTED** [review]

File: portfolio/src/main/client/src/components/sections/comments.ts [hunk 10]
```diff
@@ -20,110 +19,144 @@ export interface CommentInfo {
...
+ * @param limit The size of the page.
+ */
+async function fetchPage(
+  cursor: Cursor | null,
+  limit: number):
+  Promise<CommentPageData> {
+  const response = cursor ?
+    await fetch(`/api/comments?limit=${limit}&cursor=${cursor}`) :
+    await fetch(`/api/comments?limit=${limit}`);
+  return response.json() as Promise<CommentPageData>;
```
Comment (jpnickolas) [inline portfolio/src/main/client/src/components/sections/comments.ts]: nit: `return await response.json()`

**Review by jpnickolas — APPROVED** [review]

**Review by jessicaslaughter — APPROVED** [review]
